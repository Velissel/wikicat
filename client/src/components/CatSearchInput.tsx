import { Formik, Form, useFormikContext } from 'formik';
import React, { useCallback, useEffect } from 'react';
import Input from './Input';
import { Button, FormGroup, FormLabel, InputGroup } from 'react-bootstrap';
import { useNavigate, useSearchParams } from 'react-router-dom';

const INIT_FORM_VALUES = { searchTerm: '' };
type CatSearchInputValues = typeof INIT_FORM_VALUES;

function CatSearchTermInput() {
  const fieldName = 'searchTerm';
  const formik = useFormikContext();
  const [query] = useSearchParams();
  const q = query.get('q');

  useEffect(() => {
    if (q !== null) {
      formik.setFieldValue(fieldName, q);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [q]);

  return (
    <FormGroup className="mb-3">
      <FormLabel htmlFor={fieldName}>Search cat by breed</FormLabel>
      <InputGroup>
        <Input name={fieldName} />
        <Button type="submit">Search</Button>
      </InputGroup>
    </FormGroup>
  );
}

export default function CatSearchInput() {
  const navigate = useNavigate();
  const onSubmit = useCallback((values: CatSearchInputValues) => {
    navigate(`/search-results?q=${encodeURIComponent(values.searchTerm)}`);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Formik initialValues={INIT_FORM_VALUES} onSubmit={onSubmit}>
      <Form>
        <CatSearchTermInput />
      </Form>
    </Formik>
  );
}
