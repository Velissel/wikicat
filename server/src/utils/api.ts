import axios from 'axios';

const api = axios({
  baseURL: process.env.CAT_API,
  headers: {
    'x-api-key': process.env.CAT_API_KEY,
  },
});

export default api;
