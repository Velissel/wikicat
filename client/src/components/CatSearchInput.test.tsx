import React from 'react';
import { act, render, screen } from '@testing-library/react';
import CatSearchInput from './CatSearchInput';
import { BrowserRouter, MemoryRouter } from 'react-router-dom';
import api from '../utils/api';
import userEvent from '@testing-library/user-event';

jest.mock('../utils/api');

describe('CatSearchInput component', () => {
  beforeEach(() => {
    // @ts-ignore
    api.get.mockResolvedValue({});
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render', () => {
    const { container } = render(
      <MemoryRouter>
        <CatSearchInput />
      </MemoryRouter>,
    );
    expect(screen.getByText('Search cat by breed')).toBeInTheDocument();
    expect(
      screen.getByRole('textbox', { name: 'Search cat by breed' }),
    ).toBeInTheDocument();
    expect(screen.getByRole('button', { name: 'Search' })).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should redirect user to search page when input is submitted', async () => {
    const searchTerm = 'Turkish';
    const { container } = render(
      <BrowserRouter>
        <CatSearchInput />
      </BrowserRouter>,
    );

    expect(window.location.pathname).toEqual('/');
    expect(container).toMatchSnapshot();
    await userEvent.type(
      screen.getByRole('textbox', { name: 'Search cat by breed' }),
      searchTerm,
    );
    // eslint-disable-next-line testing-library/no-unnecessary-act
    await act(async () => {
      await userEvent.click(screen.getByRole('button', { name: 'Search' }));
    });
    expect(window.location.pathname).toEqual('/search-results');
    expect(
      screen.getByRole('textbox', { name: 'Search cat by breed' }),
    ).toHaveValue(searchTerm);
    expect(container).toMatchSnapshot();
  });
});
