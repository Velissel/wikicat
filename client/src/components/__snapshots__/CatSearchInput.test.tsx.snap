// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CatSearchInput component should redirect user to search page when input is submitted 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="searchTerm"
      >
        Search cat by breed
      </label>
      <div
        class="input-group"
      >
        <input
          class="form-control"
          id="searchTerm"
          name="searchTerm"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`CatSearchInput component should redirect user to search page when input is submitted 2`] = `
<div>
  <form
    action="#"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="searchTerm"
      >
        Search cat by breed
      </label>
      <div
        class="input-group"
      >
        <input
          class="form-control"
          id="searchTerm"
          name="searchTerm"
          type="text"
          value="Turkish"
        />
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`CatSearchInput component should render 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="searchTerm"
      >
        Search cat by breed
      </label>
      <div
        class="input-group"
      >
        <input
          class="form-control"
          id="searchTerm"
          name="searchTerm"
          type="text"
          value=""
        />
        <button
          class="btn btn-primary"
          type="submit"
        >
          Search
        </button>
      </div>
    </div>
  </form>
</div>
`;
